### PR TITLE
Orc 256 unmask range option

### PR DESCRIPTION
**Pull request by moresandeep**
This PR contains changes that enables unmasking range option for redact mask (ORC-256).

1. The react mask would accept an additional option (option #3 in this case) that has the configuration for leaving certain ranges of strings unmasked
2. The options will look like "0:4,-4:-1"
3. Range unmasking is available for 1. Long 2. Double 3. String 4. Decimal

### CODE REVIEW

**Comment by moresandeep** [comment]
Updated the PR he changes are as follows:
1. Fixed the find bugs issue.
2. Merged the feature into a single commit.

**Comment by omalley** [comment]
I think we should change the processing for numerics (when there is unmasked ranges) to be:

unmasked number -> string -> mask as string -> masked number

**Comment by moresandeep** [comment]
@omalley I updated the PR with your suggestions, thanks for the review !

File: java/core/src/java/org/apache/orc/impl/mask/RedactMaskFactory.java [hunk 1]
```diff
@@ -114,6 +120,10 @@
   private final boolean maskDate;
   private final boolean maskTimestamp;
 
+  // index tuples that are not to be masked
+  private final SortedMap<Integer,Integer> unmaskIndexRanges = Collections.synchronizedSortedMap(new TreeMap());
```
[the remark this answers is not in the document]
Comment (moresandeep) [reply java/core/src/java/org/apache/orc/impl/mask/RedactMaskFactory.java]: Hello @xndai,
Thanks for the review, I was trying to be cautious, but I can get rid of it.

File: java/core/src/java/org/apache/orc/impl/mask/RedactMaskFactory.java [hunk 2]
```diff
@@ -245,8 +271,8 @@ public void maskData(ColumnVector original, ColumnVector masked, int start,
         target.isNull[0] = source.isNull[0];
       } else {
         for(int r = start; r < start + length; ++r) {
-          target.vector[r] = maskLong(source.vector[r]) & mask;
-          target.isNull[r] = source.isNull[r];
+            target.vector[r] = maskLong(source.vector[r]) & mask;
```
[the remark this answers is not in the document]
Comment (moresandeep) [reply java/core/src/java/org/apache/orc/impl/mask/RedactMaskFactory.java]: Sure

File: java/core/src/java/org/apache/orc/impl/mask/RedactMaskFactory.java [hunk 3]
```diff
@@ -619,7 +646,7 @@ public double maskDouble(double value) {
     } else if (posn < 0) {
       posn = -posn -2;
     }
-    return DIGIT_REPLACEMENT * base * DOUBLE_POWER_10[posn];
+    return unmaskRangeDoubleValue(value,DIGIT_REPLACEMENT * base * DOUBLE_POWER_10[posn]);
```
[the remark this answers is not in the document]
Comment (moresandeep) [reply java/core/src/java/org/apache/orc/impl/mask/RedactMaskFactory.java]: ok, will do.

File: java/core/src/test/org/apache/orc/impl/mask/TestUnmaskRange.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    mask = new RedactMaskFactory("Xx7", "", "-3:-1, 0:3");
+    assertEquals(new HiveDecimalWritable("123477777.777654"),
+        mask.maskDecimal(new HiveDecimalWritable("123456789.987654")));
+
+    // test removal of leading and  trailing zeros.
+    /*
+    assertEquals(new HiveDecimalWritable("777777777777777777.7777"),
+        mask.maskDecimal(new HiveDecimalWritable("0123456789123456789.01230")));
+        */
+
```
[the remark this answers is not in the document]
Comment (moresandeep) [reply java/core/src/test/org/apache/orc/impl/mask/TestUnmaskRange.java]: ok, will do.

**Comment by moresandeep** [comment]
@xndai  @omalley I updated the PR with the suggested changes, let me know if you have any questions.

**Pull request closed without merging**